### DISCUSSION TITLE
Documentation on setting up multiple instances

### DISCUSSION THREAD

**Discussion by alexnederlof** [open-webui/open-webui#8186] [Issues]
# Conclusion of discussion below
Here's the list to run multiple instances (assuming Kubernetes):

- Enable S3, or use shared PVC
- Enable Redis and websockets
- Create a shared WEBUI_SECRET_KEY
- Specify a RAG database that is _not_ the default Chroma. That one is local disk.

As of the latest, dependencies for tools are not installed across nodes, so you can't use that functionality if you're importing dependencies.

# Original post

I've been loving this product and deploying it at my company. I had to figure a few things out to be able to have multiple instances so I can scale horizontally. I'd like to contribute to the documentation on this topic, and I think I ran into some issues. Starting this discussion for visibility to others trying the same, and to get direction on what to add to the docs, and potentially identify an issue with functions and tools. 

There's three main things I ran into

# Load balancing / session management
You need sticky web session enabled on your load balancer, because the session assumes you're talking to the same host. I'm also reading something about [Redis session managemenet](https://github.com/open-webui/docs/blob/main/docs/tutorials/integrations/redis.md) - but I'm not sure if that's another way to fix the problem, or just a way to run websockets. 

Q: what's the recommended path here that we should document?

**Update**:
You need Redis enabled to make web sockets work. In that setup, you do not need sticky sessions. 

# Assumptions around local disk
You need to use a shared disk (what i'm doing is EFS), or use  [S3 as the storage provider](https://github.com/open-webui/docs/blob/2136710910bcc70d22861c02260d5140334e26cd/docs/getting-started/advanced-topics/env-configuration.md?plain=1#L1932)

You need to deploy Chroma separately, otherwise you run into [a weird locking issue](https://github.com/open-webui/open-webui/discussions/7898) because it does not support multiple nodes reading from a disk (assuming you have a shared disk like I have through AWS EFS). It's easy enough to deploy a single Chroma instance, if you're on Kubernetes. Maybe we can add it to the helm chart?

Alternatively, you can use PGVector as the vector storage. Don't forget to set vector size https://github.com/open-webui/open-webui/blob/main/backend/open_webui/config.py#L1214

Q: Do we want to add a Chroma deployment to the helm chart? Maybe enabled it by default when someone sets `replica > 1 `?

# Instances don't talk to each other
As far as I can see, the Instances have no to communicate amongst each other. Because of that, I think the detection and installation of requirements for function or tool, only runs on the Instance that receives the HTTP request. When another instance then tries to use the tool or function, it crashes because it's missing the requirements. 

(disclaimer, I lasted tested this on 0.4.8, have not tried on the latest)

Q: Is this a known bug? And assuming we don't want a full fledged pub/sub system between nodes, maybe we can add some functionality that stores locally if it's already installed a plugin on the instance and otherwise does that before executing it?

**Comment by alexnederlof** [comment]
lol this project moves so fast, I just noticed I can also do Postgres as the backend for the RAG. Awesome stuff. I think that just means the docs should mention you should not use Chroma on disk if you plan to do multi instance

**Comment by ajikmr** [comment]
Hi,
I am hosting it on AWS behind a load balancer on EC2 (vm). Google is used for oAuth.

When single vm is used it works fine, but when mutiple VMs are used behind load balancer it does not work. It will be good to have a sample **production version** of docker compose.

**Reply by alexnederlof** [reply]
Do you have sticky sessions enabled?

**Reply by ajikmr** [reply]
I would appreciate any help in this regard. By sticky session, do you mean websocket? I have following in the .env for docker as suggested in documentation.

ENABLE_WEBSOCKET_SUPPORT="True"
WEBSOCKET_MANAGER="redis"
WEBSOCKET_REDIS_URL="redis://redis:6379/1"

I would appreciate some help in regard to sticky session. Is there documentation (or sample docker-compose) regarding it?

**Reply by ajikmr** [reply]
I am running multiple instance of open-webui behind a load balancer, and the setup  is backed by a separate postgres db. All open-webui instances have access to postgres db. 

It gives error for multple instances of open-webui, but for single instance it works fine.

**Reply by alexnederlof** [reply]
Sticky sessions you need to setup on your LB https://duckduckgo.com/?q=aws+alb+sticky+session&ia=web. You need a setup where it installs a cookie and routes based on that

**Reply by jk-f5** [reply]
As of #7887 sticky sessions aren't required for multiple replicas. Redis is required to track websocket sessions across replicas. I would also suggest using something other than Sqlite (the default open webui backing database) like Postgres.

I run multiple replicas without sticky sessions with no issues.

**Reply by ajikmr** [reply]
Ok. Thanks for your tips. Lets me try these and see what works.

**Reply by ajikmr** [reply]
Hi,
It is working for me now. The configuration working for me is to use socket with redis, along with stickiness in aws load balancer. And using postgres as db.

websocket should be working without the need for stickiness in aws load balancer. I will check why why I need load balancer stickiness. (May be websocket using redis is not set up properly??)

**Reply by jk-f5** [reply]
Turn off sticky sessions and check your browser network activity through "developer tools". Are you seeing repeated hits to the `socket.io` path with the `polling` option? What are you seeing in terms of failed requests? Are you sure you're not running an older version of the web UI? (ie did you update open webui but not clear your cache and refresh any existing browser tabs that already had open webui running?)

**Reply by alexnederlof** [reply]
Cool, I'll try the redis setup then

**Reply by alexnederlof** [reply]
Can confirm switching to Redis fixed all network problems (folks getting 401s, or unresponsive UI). Sticky sesssions no longer needed

**Reply by rragundez** [reply]
do i need a separate Redis DB or having this config below is enough?

ENABLE_WEBSOCKET_SUPPORT="True"
WEBSOCKET_MANAGER="redis"
WEBSOCKET_REDIS_URL="redis://redis:6379/1"

**Reply by jk-f5** [reply]
Answered this same question in #5109

**Comment by df-cgdm** [comment]
I'm running open-webui version 0.5.7, I have redis configured, an external Postgresql database (used both for user and vector db), S3 for file and I'm using oauth authentication.

As soon I set the replicas to more than one, I'm getting a message "The email or password provided is incorrect. Please check for typos and try logging in again."
On the backend log I have the following message:
"OAuth callback error: mismatching_state: CSRF Warning! State not equal in request and response"

Is there any other configuration to be done?

**Reply by df-cgdm** [reply]
Finallly find the issue. 
The problem was related to the WEBUI_SECRET_KEY which must be shared between instance.
As I'm not using anymore pvc I need have to set the environment variable

**Reply by alexnederlof** [reply]
ah yes, good one. I'll add it to the list
